src/mds/MDSAuthCaps.cc

```diff
@@ -410,7 +410,11 @@ bool MDSAuthCaps::merge_one_cap_grant(MDSCapGrant ng)
       // fsname and path match but value of root_squash is different. update
       // its value.
       if (g.match.root_squash != ng.match.root_squash) {
-	g.match.root_squash = ng.match.root_squash;
+	// "fs authorize" command is not allowed to deduct caps. so, we can add
+	// but not remove root_squash from MDS auth caps.
+	if (g.match.root_squash == false) {
+	  g.match.root_squash = ng.match.root_squash;
+	}
```

Comment:
@rishabh-d-dave Could you point me the doc where says that the `fs authorize` won't allow us to change the caps ?

What I get from the doc https://docs.ceph.com/en/reef/cephfs/client-auth/ is that we could, but not always:

```
Capabilities that have been issued to a client can not be removed by running fs authorize again. For example, if a client capability has root_squash applied on a certain CephFS, running fs authorize again for the same CephFS but without root_squash will not lead to any update and the client caps will remain unchanged:
```
That means if the `caps` are not issued to any client, the change should succeed ?

Reply:
> @rishabh-d-dave Could you point me the doc where says that the `fs authorize` won't allow us to change the caps ?

Yes, you are quoted the exact part of CephFS doc that says o.

> 
> What I get from the doc https://docs.ceph.com/en/reef/cephfs/client-auth/ is that we could, but not always:
> 
> ```
> Capabilities that have been issued to a client can not be removed by running fs authorize again. For example, if a client capability has root_squash applied on a certain CephFS, running fs authorize again for the same CephFS but without root_squash will not lead to any update and the client caps will remain unchanged:
> ```
> 

Yes, that's the exact part.

> That means if the `caps` are not issued to any client, the change should succeed ?

It has nothing to do with clients. What it means is we can add new caps to client keyring using `fs authorize` but not remove caps using it. You can read this section too - https://docs.ceph.com/en/reef/cephfs/client-auth/#changing-rw-permissions-in-caps

Reply:
`Capabilities can be modified by running fs authorize only in the case when read/write permissions must be changed.`

I think this just means we can only change the `read/write` permissions, right ?

Reply:
> I think this just means we can only change the read/write permissions, right ?

Yes.

> If my understanding is correct then we can add or deduce the rw caps I think. And then the comments in you change maybe not proper ?

The comment is on for `root_squash` not `rw`.

Reply:
This sounds fine.

Do we have tests that cover `rw` part @rishabh-d-dave? If note, we should add those (in a separate change).

Reply:
@rishabh-d-dave ping?

Reply:
@rishabh-d-dave This comment is pending an update. PTAL to avoid merge delays.

Reply:
Hi Venky. There's no change needed to be made based on this thread AFAIU.

Reply:
> Hi Venky. There's no change needed to be made based on this thread AFAIU.

ping

Reply:
OK, so the comment could be misread a bit like @lxbsz said.

```
// "fs authorize" command is not allowed to deduct caps. so, we can add but not remove root_squash from MDS auth caps.
```

If `rw` auth caps can be reduced to say `r` then please mention that in the comment.

Reply:
This comment meant only for root_squash as it lies on top of code for root_squash. Adding a mention here for permissions won't make sense IMO.

For permissions a comment is already present. See: https://github.com/ceph/ceph/blob/main/src/mds/MDSAuthCaps.cc#L404-L405

Reply:
> "fs authorize" command is not allowed to deduct caps.

Is pretty generic and therefore the suggestion.

Reply:
no worries - doesn't block merge.